Custom local variables in plugins

Goldorion opened the pull request:
The same PR as #1054, but it doesn't contain the problems after merging the master branch.
Requested changes from the original PR have also been applied.

KlemenDEV commented on src/test/java/net/mcreator/integration/generator/GTModElements.java:
```diff
@@ -96,7 +96,7 @@ public static void runTest(Logger LOG, String generatorName, Random random, Work
 					ModElement modElement1 = workspace.getGenerator().getModElementThisFileBelongsTo(modElementFile);
 					if (!modElement.equals(modElement1))
-						fail("Filed to properly determine file ownership for mod element type: " + modElement.getType()
+						fail("Filed to properly determine file ownership for mod element type: " + modElement.getName()
```
Why did you replace getType with getName here?

Goldorion replied:
Oops, this is due to the refractoring.

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElement.java:
```diff
@@ -18,10 +18,16 @@
 
 	private String name;
-	private VariableElementType type;
+	private String type;
```
What happens with existing variables in workspace due to this change?

Goldorion replied:
Current variables are saved with a String, so I made this change to be compatible with old workspaces.

KlemenDEV replied:
But they are probably saved as uppercase, not lowercase string like now? Also please don't resolve yourself as I then don't notice your comments. Thanks!

Goldorion replied:
No, both are lower case strings.
Before/Now:
![image](https://user-images.githubusercontent.com/38427877/116956049-49364d00-ac62-11eb-9964-7a7fdbf4a67a.png)

This PR:
![image](https://user-images.githubusercontent.com/38427877/116956070-58b59600-ac62-11eb-986e-11bd97f7354d.png)

KlemenDEV replied:
Oh, you are right, my bad. Fine then ;)

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElement.java:
```diff
@@ -38,10 +44,32 @@
+	}
+
+	public static VariableElementType getVariableFromType(String type) {
+		for (VariableElementType var : variables.keySet()) {
```
Still using var

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/NewVariableDialog.java:
```diff
@@ -41,7 +42,7 @@ public static VariableElement showNewVariableDialog(JFrame frame, boolean showSc
 		textField.setValidator(variableNameValidator);
 
-		JComboBox<VariableElementType> type = new JComboBox<>(supportedTypes);
+		JComboBox<String> type = new JComboBox<>(supportedTypes);
```
I suggest you keep using variable object here and not string, applies to most other places too. It should implement toString to provide string value, but I see no need in refactoring to String as you did in most places.

KlemenDEV commented on src/test/java/net/mcreator/integration/generator/GTModElements.java:
```diff
@@ -110,7 +110,7 @@ public static void runTest(Logger LOG, String generatorName, Random random, Work
 					if (modElement
 							.equals(modElement1)) // if now ownership can still be found, this means some files were not properly removed
-						fail("Filed to properly delete file of mod element type: " + modElement.getType()
+						fail("Filed to properly delete file of mod element type: " + modElement.getName()
```
You should print type, not name here

KlemenDEV commented on plugins/generator-1.16.5/forge-1.16.5/templates/modbase/variableslist.java.ftl:
```diff
@@ -355,16 +355,11 @@ public class ${JavaModName}Variables {
-			</#if>
-		</#list>
-		if(!event.isWasDeath()) {
+		if(event.isWasDeath()) {
```
This part should not be altered as these changes do not fall under this PR.

Goldorion replied:
Weird... I never changed something at this place. This is the first I see this event. :/

KlemenDEV replied:
Well revert that part so no changes apply here

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/blocks/ProcedureRetvalBlock.java:
```diff
@@ -24,33 +24,27 @@
-			return;
+	public ProcedureRetvalBlock() {
+		for (VariableElementType var : VariableElement.getVariables()) {
+			names.add("procedure_retval_" + var.getBlockName());
```
Make array at this point and store array in the object field. Same for other IBlockGenerators

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElementType.java:
```diff
@@ -22,55 +22,69 @@
-		this.color = color;
+	private String name;
+	private String color;
+	//Use for compatibility with old workspaces
```
Could you explain why is this here and what is its purpose? And are all these fields for compatibility or only this one and blockName and how is this used?

Goldorion replied:
Fields used for compatibility are only `dependencyType` and `blockName`. The other fields are needed by all variables as they can't use the `type` field. I will change the order to make that more understandable.

KlemenDEV replied:
Please explain what you mean by compatibility and what value do these hold and why we need them as I can't fully understand them

Goldorion replied:
Concerning `dependencyType`, in the mappings, you decided to name the dependency used for the LOGIC variables `boolean`, so I couldn't change it for `logic` without breaking workspaces. I decided to add `dependencyType` to solve this problem.
Concerning `blockName`, it's the same problem, except you named all String procedure blocks using `text` instead of String. So, I had to add this new field, so String variables and variable blocks will not break old workspaces.

They have been simply added to fix these two problems because, for all other variables, the `name` field will be used by default.

Do you understand know what I mean by compatibility with old workspaces and why they have been added?

KlemenDEV replied:
I do. But this makes code very hard to read. I suggest avoiding legacy code and adapting workspace format for old workspaces instead. Remove legacy code and update workspaces when reading them.

This will also assure names will be unified from now on to fix my mistake with different names.

Check how I added backwards compatibility when I added multiple sounds support:

1. https://github.com/MCreator/MCreator/blob/master/src/main/java/net/mcreator/workspace/WorkspaceFileManager.java#L48
2. https://github.com/MCreator/MCreator/blob/852200163b48d9d8184953a4755f1e79ede550ba/src/main/java/net/mcreator/workspace/elements/SoundElement.java#L89

Goldorion replied:
For the `blockName`, I made a converter as it was inside procedure mod elements and for the dependencyType, I simply changed everything inside the code. When you will have read this message, you can resolve the conversation.

KlemenDEV replied:
Ok. I will do review of the change in the upcoming days when I get a bit larger chunk of time :)

KlemenDEV reviewed (changes requested):
Some suggested changes and some questions I have.

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElement.java:
```diff
@@ -38,10 +43,32 @@
+		return variables.keySet();
+	}
+
+	public void setType(String type) {
```
I think this method should take VariableElementType and then convert to string inside here.

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElement.java:
```diff
@@ -38,10 +43,32 @@
+		return variables.keySet();
+	}
+
+	public static void addVariable(VariableElementType var) {
```
I don't understand what this does. addVariable but it takes variable type, what does this system do and why is it needed?

Goldorion replied:
This method is called every time a variable is loaded and can be used. It will add the variable gave in the `VARIABLE_CACHE` map variable. Then, this `VARIABLE_CACHE` map variable can use in the other getters of the class. I renamed it for `addVariableTypeToCache`.
It only takes a `VariableElementType` as a parameter because we can get its name directly with the VariableElementType.

KlemenDEV reviewed (changes requested):
Two more things that caught my attention

KlemenDEV commented on plugins/generator-1.15.2/forge-1.15.2/mappings/types.yaml:
```diff
@@ -5,7 +5,7 @@ entity: Entity
 itemstack: ItemStack
 map: HashMap
 string: String
-boolean: boolean
```
Check procedure dependencies, they use "boolean" right now so you need to change it there too, and on other places where it may need replacement now (eg. ParticleGUI class, Dependency class). Please test these changes well too.

Goldorion replied:
Oh, I thought I already changed procedure blocks. I will check that.

KlemenDEV reviewed (changes requested):
> java.lang.RuntimeException: Failed to load any translation for key: blockly.block.custom_dependency_string

Seems there are some missing translation keys

Goldorion commented:
> > java.lang.RuntimeException: Failed to load any translation for key: blockly.block.custom_dependency_string
> 
> Seems there are some missing translation keys

I didn’t update existing translation keys for them.

KlemenDEV commented:
This PR needs fixing of some merge conflicts that occured

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/blocks/GetVariableBlock.java:
```diff
@@ -27,35 +27,24 @@
-		default:
-			return;
+	public GetVariableBlock() {
+		for (VariableElementType var : VariableElement.getVariables()) {
```
You are still using var here

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElement.java:
```diff
@@ -38,11 +43,28 @@
+		return VARIABLE_CACHE.keySet();
+	}
+
+	public static void addVariableTypeToCache(VariableElementType var) {
```
I see what you mean here. I think you could move this to variableloader and call this VARIABLE_LIST, not VARIABLE_CACHE

KlemenDEV commented on src/main/java/net/mcreator/ui/minecraft/ProcedureSelector.java:
```diff
@@ -116,7 +117,7 @@ public ProcedureSelector(@Nullable IHelpContext helpContext, MCreator mcreator,
 			defaultName = "(always)";
 			setBorder(BorderFactory
-					.createLineBorder(new Dependency("", VariableElementType.LOGIC.toDependencyType()).getColor()));
+					.createLineBorder(new Dependency("", VariableElementType.LOGIC.getName()).getColor()));
```
Consider checking if returnType is not null in separate if and setting border there, so all dependency type colors are covered

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElementType.java:
```diff
@@ -22,55 +22,48 @@
 
-	private final int color;
+	//Define each global variables. The instantiation is made in VariableElementTypeLoader.
+	public static VariableElementType STRING;
```
I would move these static hardcodes to valriabeloader class too

KlemenDEV commented on src/main/java/net/mcreator/blockly/BlocklyBlockUtil.java:
```diff
@@ -118,4 +124,73 @@ public static Color getBlockColorFromHUE(int hue) {
+			return "\"%{BKY_LOGIC_HUE}\"";
+
+		//Custom colors
+		float[] hsbValues = new float[3];
```
I think you can merge this into one line and put null as argument according to https://docs.oracle.com/javase/7/docs/api/java/awt/Color.html#RGBtoHSB(int,%20int,%20int,%20float[])

KlemenDEV commented on plugins/mcreator-core/variables/itemstack.json:
```diff
@@ -0,0 +1,7 @@
+{
+	"name": "itemstack",
+	"color": "0x996069",
```
I just thought that maybe we could just use an integer here similar to blockly (hue only), as all colors use blockly lightness and saturation so we only truly need hue here. This also would allow matching colors with procedure blocks more easily as they only specify hue too.

KlemenDEV commented on src/main/java/net/mcreator/ui/blockly/BlocklyPanel.java:
```diff
@@ -143,6 +145,9 @@ public BlocklyPanel(MCreator mcreator) {
 							.replace("@RESOURCES_PATH", resDir));
 
+					//JS code generation for custom variables
+					webEngine.executeScript(VariableElementTypeLoader.INSTANCE.JS_CACHE.toString());
```
I think JS_CACHE should be string so you don't need to use toString here

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -111,7 +112,7 @@ public static void fillWorkspaceWithTestData(Workspace workspace) {
 		sampleVariable1.setName("test");
 		sampleVariable1.setValue("true");
-		sampleVariable1.setType(VariableElementType.LOGIC);
+		sampleVariable1.setType(VariableElementTypeLoader.LOGIC);
```
Maybe these should be moved to subclass for code clarity, so calls like this would look like:

`VariableElementTypeLoader.BuiltInTypes.LOGIC`

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElementTypeLoader.java:
```diff
@@ -0,0 +1,108 @@
+		}
+	}
+
+	public static VariableElementType getVariableFromType(String type) {
```
Rename to something like `getVariableTypeFromString`

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/BlocklyVariables.java:
```diff
@@ -43,9 +43,10 @@
-				if (JavaKeywordsMap.VARIABLE_TYPES.get(type) != null && name != null) {
-					generator.append(JavaKeywordsMap.VARIABLE_TYPES.get(type)[0]).append(" ").append(name).append(" = ")
-							.append(JavaKeywordsMap.VARIABLE_TYPES.get(type)[1]).append(";\n");
+				VariableElementType var = VariableElementTypeLoader.getVariableFromType(type);
```
You are still using var here

KlemenDEV commented on src/main/java/net/mcreator/workspace/elements/VariableElementType.java:
```diff
@@ -22,55 +22,42 @@
+	}
 
-		return null;
+	public String getJavaClass() {
```
Looking again this may not be even needed as you already have `getJavaType`, or is this one different. I think `getJavaType` could be used in all cases as it also uses mapping system

Goldorion replied:
You're right. The `getJavaType` method using mappings is the only one needed.

KlemenDEV commented on src/main/java/net/mcreator/ui/modgui/GameRuleGUI.java:
```diff
@@ -175,7 +175,7 @@ private void updateDefaultValueUI() {
 		super.afterGeneratableElementStored();
 		modElement.clearMetadata();
-		modElement.putMetadata("type", "Number".equals(gameruleType.getSelectedItem()) ? "number" : "boolean");
+		modElement.putMetadata("type", "Number".equals(gameruleType.getSelectedItem()) ? "number" : "logic");
```
You will need to make updater to update metadata of existing mod elements from boolean to logic

Also maybe consider using eg. `VariableElementTypeLoader.LOGIC.getName()` here to re-use new united names

KlemenDEV commented on src/main/java/net/mcreator/ui/blockly/BlocklyJavascriptTemplates.java:
```diff
@@ -0,0 +1,77 @@
+
+public class BlocklyJavascriptTemplates {
+
+	public static String variableListExtension(VariableElementType var) {
```
Still using var here a lot

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/blocks/ReturnBlock.java:
```diff
@@ -25,31 +25,26 @@
-			returnType = VariableElementType.ITEMSTACK;
-			break;
+	public ReturnBlock() {
+		Set<String> temp = new HashSet<>();
```
You should use stream with .map and .collect and array collector here for cleaner code. Applies to all IBlockGenerators

Goldorion commented:
Klemen, why tests don't have `x`, `y` and `z`? When I test some blocks failing in tests, everything works properly.

KlemenDEV commented:
Hmm that is strange, I can't see problem in changed files, but there must be something there. Try running with debugger or printing state at different stages inside procedures tests to see what part fails.

Goldorion commented:
> 
> 
> Hmm that is strange, I can't see problem in changed files, but there must be something there. Try running with debugger or printing state at different stages inside procedures tests to see what part fails.

@KlemenDEV I think I found the cause. I forgot to re-add the beginning of the block names for the 5 generated blocks. I changed again the code, but if you still don't like this new code, do the code like you want because I changed the code 10 times and you never liked it. I'm out of ideas and motivation for them. I know you want a good code, but sometimes it's a bit much.
Anyways, tests should be ok now and you can do your final review. ;)

KlemenDEV commented:
> I changed again the code, but if you still don't like this new code, do the code like you want because I changed the code 10 times and you never liked it. I'm out of ideas and motivation for them. I know you want a good code, but sometimes it's a bit much.

I started to review it and found multiple problems already. Not related to good code but to functional code (eg. wrong converter number). If I fix this myself, this does not differentiate much from feature requests if I need to spend a day or two on a code of PR, if you see my point. Also if I fix your problems, you will never learn about them. I know my comments can be annoying. Just so we are clear why I leave comments instead of fixing it myself, even if it seems annoying at the first look :)

KlemenDEV commented on src/main/java/net/mcreator/ui/workspace/WorkspacePanelVariables.java:
```diff
@@ -18,6 +18,7 @@
 
 package net.mcreator.ui.workspace;
```
Clicking dropdowns in this panel of variable type throws an exception. Check this, fix, and test.

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/NewVariableDialog.java:
```diff
@@ -66,9 +68,10 @@ public static VariableElement showNewVariableDialog(JFrame frame, boolean showSc
 			element.setName(Transliteration.transliterateString(textField.getText()));
 			element.setType((VariableElementType) type.getSelectedItem());
-			element.setValue(VariableElement.getDefaultValueForType((VariableElementType) type.getSelectedItem()));
+			element.setValue(Objects.requireNonNull(variable).getDefaultValue());
```
I would check if variable is null and in this case return null from this method (as if frame is closed)

KlemenDEV commented on src/main/java/net/mcreator/ui/blockly/BlocklyPanel.java:
```diff
@@ -223,12 +228,12 @@ public void removeLocalVariable(String name) {
 				VariableElement element = new VariableElement();
 				element.setName(vardata[0]);
-				element.setType(BlocklyVariables.getMCreatorVariableTypeFromBlocklyVariableType(vardata[1]));
+				element.setType(Objects.requireNonNull(BlocklyVariables.getVarElementTypeFromBlocklyType(vardata[1])));
```
It can happen plugin is removed and the variable no longer exists. This will then crash. Do null check and skip that variable (continue statement)

KlemenDEV commented on src/main/java/net/mcreator/ui/blockly/BlocklyPanel.java:
```diff
@@ -223,12 +228,12 @@ public void removeLocalVariable(String name) {
 		String[] vars = query.split(":");
-		for (String var : vars) {
-			String[] vardata = var.split(";");
+		for (String varName : vars) {
```
This is not varName but var (name:type) so this new naming is misleading

KlemenDEV reviewed (changes requested):
I did my changes (fixes) myself, but some should be fixed by you I think (see my comment).

KlemenDEV reviewed (commented):
I did all changes that seemed necessary. @Goldorion please test on some of your workspaces as I did some changes, also asking @Max094Reikeb and @AlsoSomeoneElse to take a look as this PR can easily break a lot of things if we miss any details out. I am just starting doing some tests myself too.

AlsoSomeoneElse commented on plugins/mcreator-core/variables/logic.json:
```diff
@@ -0,0 +1,7 @@
+{
+	"name": "logic",
+	"dependencyType": "boolean",
```
I don't think this is actually used anywhere, because I couldn't find any other instance of `dependencyType`

AlsoSomeoneElse commented on plugins/generator-1.15.2/forge-1.15.2/procedures/_set_variable.java.ftl:
```diff
@@ -72,13 +72,13 @@
-    <#elseif type == "ITEMSTACK">
+    <#elseif type == "itemstack">
         ${name} = ${mappedMCItemToItemStackCode(value, 1)};
     </#if>
```
The "Set variable" block doesn't actually generate any code for other variable types (same for 1.16)

KlemenDEV replied:
Keeping unresolved so I don't forget to test this

KlemenDEV replied:
Yup this works now :)

Goldorion commented:
When I fixed problems SomeoneElse reported, I noticed that only built-in variable types could have a color for the dependency type. I added a small condition checking if the type (of `Dependency#getColor`) is a loaded variable type and if true, returning the HUE color defined in the JSON file. This way, all custom variable types can have a color (instead of white), and built-in variable types have the same color everywhere.

Another I made is now, the set variable block now requires the tye of the variable inside the `types.yaml` mapping file (`variableName: javaType`), so the generated code can properly cast the value.

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/blocks/SetVariableBlock.java:
```diff
@@ -35,25 +38,21 @@
+		String javaType = new Dependency("",
+				VariableElementTypeLoader.INSTANCE.getVariableTypeFromString(type).getName())
+				.getType(master.getWorkspace());
+		if (javaType.equals("Object")) {
```
If object is used, this is fine as one can always assign anything to an object. So I don't think this is 100% needed. Bigger problem is with get block where object will be used in eg. integer context. But this can't really happen so this check can probably be safely skipped.

KlemenDEV reviewed (approved):
@Goldorion I have moved default variable values to the mappings system as some generators may use different default values. Waiting for some tests from you and other contributors and then it is ready to merge I think.

Goldorion commented:
> @Goldorion I have moved default variable values to the mappings system as some generators may use different default values. Waiting for some tests from you and other contributors and then it is ready to merge I think.

I will test when I get back home.

AlsoSomeoneElse commented on src/main/java/net/mcreator/ui/workspace/WorkspacePanelVariables.java:
```diff
@@ -125,11 +127,11 @@
 					};
 				} else if (modelColumn == 3) {
 					VariableElementType data = (VariableElementType) elements.getValueAt(row, 1);
```
This line causes some issues because it tries to cast a string into a VariableElementType, I think it should be `VariableElementType data = VariableElementTypeLoader.INSTANCE.getVariableTypeFromString((String) elements.getValueAt(row, 1));`

Goldorion replied:
You're right. I forgot this one when I fixed the panel.

AlsoSomeoneElse commented on src/main/java/net/mcreator/ui/workspace/WorkspacePanelVariables.java:
```diff
@@ -78,12 +80,12 @@
-							VariableElementType type = (VariableElementType) getValueAt(row, column);
-							if (type == VariableElementType.NUMBER) {
+							VariableElementType type = VariableElementTypeLoader.INSTANCE.getVariableTypeFromString((String) getValueAt(row, column));
+							if (type == VariableElementTypeLoader.BuiltInTypes.NUMBER) {
```
I think this part could be edited so that it sets the variable default value instead of checking each type

Goldorion replied:
No, because ItemStack (and future variable types) use something that the majority of users won't understand. For example, if I add a Date var type, the default value will be `new Date()` and you can't really change this for a value like a number or a text. Also, someone new to MCreator or who doesn't have Java knowledge will not understand what it is. I prefer to only keep these 3 var types, so the UI is clean and understandable by everyone.
